oxidation/libparsec/crates/types/src/addr.rs

```diff
@@ -157,42 +188,27 @@ struct BaseBackendAddr {
 }
 
 impl BaseBackendAddr {
-    fn from_url(parsed: &Url, pairs: &url::form_urlencoded::Parse) -> Result<Self, AddrError> {
-        if parsed.scheme() != PARSEC_SCHEME {
-            lazy_static! {
-                static ref SCHEME_ERROR_MSG: String =
-                    format!("Must start with `{}://`", PARSEC_SCHEME);
-            }
-            return Err(&SCHEME_ERROR_MSG);
-        }
-
+    fn from_url(parsed: &ParsecUrlAsHTTPScheme) -> Result<Self, AddrError> {
         // Use the same error message than for `Url::parse` because
         // `Url::parse` considers port with no hostname (e.g. `http://:8080`)
         // invalid and we want to stay consistent to simplify testing
-        let hostname = parsed.host_str().ok_or("Invalid URL")?;
-
-        let mut no_ssl_queries = pairs.filter(|(k, _)| k == "no_ssl");
-        let use_ssl = match no_ssl_queries.next() {
-            None => true,
-            Some((_, value)) => match value.to_lowercase().as_str() {
-                // param is no_ssl, but we store use_ssl (so invert the values)
-                "false" => true,
-                "true" => false,
-                _ => {
-                    return Err("Invalid `no_ssl` param value (must be true or false)");
-                }
-            },
+        let hostname = parsed.0.host_str().ok_or("Invalid URL")?;
+
+        // `ParsecUrlAsHTTPScheme`'s creation has consumed the `no_ssl` param to
+        // determine if it should be an http or https
+        let use_ssl = match parsed.0.scheme() {
+            "http" => false,
+            "https" => true,
+            _ => unreachable!("scheme can only be http or https"),
         };
-        if no_ssl_queries.next().is_some() {
-            return Err("Multiple values for param `no_ssl`");
-        }
 
         let default_port = if use_ssl {
             PARSEC_SSL_DEFAULT_PORT
         } else {
             PARSEC_NO_SSL_DEFAULT_PORT
         };
         let port = parsed
+            .0
             .port()
             .and_then(|p| if p == default_port { None } else { Some(p) });
```

Comment:
Could this logic moved to `ParsecUrlAsHTTPScheme` ?

Could be logical that `ParsecUrlAsHTTPScheme` provides methods to get the `hostname`, `port` & `use_ssl`

Reply:
I'd rather have it here and keep `ParsecUrlAsHTTPScheme` as simple as possible: it is just a simple wrapper to indicate the Url object is a bit hacky, but in a perfect world it would be needed and a parsec:// url would be passed here